risingwavelabs/risingwave
# feat(stream): Add recovery feature for nats source

I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- Summarize your change: Add nats recovery mechanism.
- How does this PR work? Need a brief introduction for the changed logic (optional)
    - In [nats jetstream message](https://docs.rs/async-nats/latest/async_nats/jetstream/message/struct.Message.html#method.info), it has a [method](https://docs.rs/async-nats/latest/async_nats/jetstream/message/struct.Message.html#method.info) to get the message [info](https://docs.rs/async-nats/latest/async_nats/jetstream/message/struct.Info.html) which we can get the sequence id. It has `consumer sequence` and `stream sequence`, we use the [consumer sequence](https://docs.rs/async-nats/latest/async_nats/jetstream/message/struct.Info.html#structfield.consumer_sequence).
    - When [create consumer](https://docs.rs/async-nats/latest/async_nats/jetstream/stream/struct.Stream.html#method.get_or_create_consumer), we can specify the [config](https://docs.rs/async-nats/latest/async_nats/jetstream/consumer/pull/struct.Config.html), in [config](https://docs.rs/async-nats/latest/async_nats/jetstream/consumer/pull/struct.Config.html#structfield.deliver_policy), we can set the `deliver_policy` as [`ByStartSequence`](https://docs.rs/async-nats/latest/async_nats/jetstream/consumer/enum.DeliverPolicy.html).
    - We use the sequence number as the offset in risingwave to implement recovery feature.

- Describe clearly one logical change and avoid lazy messages:
    - in previous code, we cannot get the message when risinwgave broken down and also cannot get it after back to work. But now, when rw recovered, it will continue to consume the message sended by server during the risingwave broken down.
- Describe any limitations of the current code: Test under one cluster address. Not test under [multiple clusters](https://docs.rs/async-nats/latest/async_nats/fn.connect.html#connect-with-vec-of-serveraddr).
- I test it under 2 situation and rw can work correctly under both:
    - testcase 1:
     1. In nats, send message into subject1 from 0 to 99
     2. Check rw get data from 0 to 99
     3. Kill compute node
     4. In nats, send message into subject1 from 100 to 199
     5. Restart rw cluster
     6. Get the data from 0-199, no duplicate and no lost.
    

    - testcase 2：  provided by @tabVersion to make sure no overlap in sequence number (which is the 'offset' we used to do recovery):
     1. Open the nats, always send message to subject2 regularly and none stop.
     2. do the steps in testcase1


## Checklist

- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [x] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

officially support NATS, @yufansong please update the description here, and introduce your NATS options

demo SQL 

```sql
CREATE TABLE test
(
  a int
)
WITH (
  connector='nats',
    nats.server_url='0.0.0.0:4222',
    nats.subject='test1',
) FORMAT PLAIN ENCODE JSON;
```




</details>
